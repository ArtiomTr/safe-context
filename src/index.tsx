import React, { Context, createContext, FunctionComponent, PropsWithChildren, useContext } from 'react';
import invariant from 'tiny-invariant';

export type SafeProviderProps<T> = PropsWithChildren<{
    value: T;
}>;

export type SafeConsumerProps<T> = {
    children: (value: T) => React.ReactNode;
};

export type SafeContext<T> = {
    Provider: FunctionComponent<SafeProviderProps<T>>;
    Consumer: FunctionComponent<SafeConsumerProps<T>>;
    displayName?: string;
};

export const useSafeContext = <T,>(safeContext: SafeContext<T>): T => {
    const contextValue = useContext((safeContext as unknown as { context: Context<T | undefined> }).context);

    invariant(
        contextValue !== undefined,
        `Trying to access ${safeContext.displayName ?? 'SafeContext'} outside provider.`,
    );

    return contextValue;
};

export const createSafeContext = <T,>(): SafeContext<NonNullable<T>> => {
    const UnsafeContext = createContext<NonNullable<T> | undefined>(undefined);

    const safeContext = {
        Provider: ({ value, children }: SafeProviderProps<NonNullable<T>>) => {
            invariant(value !== undefined, 'SafeContext.Provider cannot provide `undefined` value.');

            return <UnsafeContext.Provider value={value}>{children}</UnsafeContext.Provider>;
        },
        Consumer: ({ children }: SafeConsumerProps<NonNullable<T>>) => {
            const value = useSafeContext(safeContext);

            return <React.Fragment>{children(value)}</React.Fragment>;
        },
        context: UnsafeContext,
    };

    return safeContext;
};
