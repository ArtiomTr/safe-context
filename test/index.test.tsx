import React, { Context, useContext } from 'react';
import { render, renderHook } from '@testing-library/react';

import { createSafeContext, useSafeContext } from '../src';

describe('createSafeContext', () => {
    it('should not throw', () => {
        expect(() => createSafeContext()).not.toThrow();
    });

    it('should return new context', () => {
        const context = createSafeContext();

        expect(context.Provider).toBeDefined();
        expect(context.Consumer).toBeDefined();
    });
});

describe('SafeProvider', () => {
    it('should work as normal provider', () => {
        const SampleContext = createSafeContext();

        const testValue = {};

        const { result } = renderHook(
            () => useContext((SampleContext as unknown as { context: Context<unknown> }).context),
            {
                wrapper: ({ children }) => (
                    <SampleContext.Provider value={testValue}>{children}</SampleContext.Provider>
                ),
            },
        );

        expect(result.current).toBe(testValue);
    });

    it('should throw error if value is undefined', () => {
        const SampleContext = createSafeContext();

        let error: Error | undefined = undefined;

        try {
            render(<SampleContext.Provider value={undefined} />);
        } catch (receivedError: unknown) {
            error = receivedError as Error;
        }

        expect(error).toBeDefined();
    });
});

describe('SafeConsumer', () => {
    it('should work as normal consumer', () => {
        const SampleContext = createSafeContext();

        const testValue = {};

        const consumerTester = jest.fn();

        render(
            <SampleContext.Provider value={testValue}>
                <SampleContext.Consumer>{consumerTester}</SampleContext.Consumer>
            </SampleContext.Provider>,
        );

        expect(consumerTester.mock.calls[0][0]).toBe(testValue);
    });

    it('should throw error if not wrapped into provider', () => {
        const SampleContext = createSafeContext();

        let error: Error | undefined = undefined;

        const consumerTester = jest.fn();

        try {
            render(<SampleContext.Consumer>{consumerTester}</SampleContext.Consumer>);
        } catch (receivedError: unknown) {
            error = receivedError as Error;
        }

        expect(error).toBeDefined();
        expect(consumerTester).not.toBeCalled();
    });
});

describe('useSafeContext', () => {
    it('should work as normal useContext', () => {
        const SampleContext = createSafeContext();

        const testValue = {};

        const { result } = renderHook(() => useSafeContext(SampleContext), {
            wrapper: ({ children }) => <SampleContext.Provider value={testValue}>{children}</SampleContext.Provider>,
        });

        expect(result.current).toBe(testValue);
    });

    it('should throw error if not wrapped into provider', () => {
        const SampleContext = createSafeContext();
        SampleContext.displayName = 'SampleContext';

        let error: Error | undefined = undefined;

        try {
            renderHook(() => useSafeContext(SampleContext));
        } catch (receivedError: unknown) {
            error = receivedError as Error;
        }

        expect(error).toBeDefined();
        expect(error?.message).toMatchSnapshot();
    });

    it('should throw error if not wrapped into provider (no display name)', () => {
        const SampleContext = createSafeContext();
        SampleContext.displayName = undefined;

        let error: Error | undefined = undefined;

        try {
            renderHook(() => useSafeContext(SampleContext));
        } catch (receivedError: unknown) {
            error = receivedError as Error;
        }

        expect(error).toBeDefined();
        expect(error?.message).toMatchSnapshot();
    });
});
