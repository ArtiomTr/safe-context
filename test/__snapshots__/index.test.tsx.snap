// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSafeContext should throw error if not wrapped into provider (no display name) 1`] = `"Invariant failed: Trying to access SafeContext outside provider."`;

exports[`useSafeContext should throw error if not wrapped into provider 1`] = `"Invariant failed: Trying to access SampleContext outside provider."`;
